src/gt4py/next/iterator/type_system/inference.py

```diff
@@ -0,0 +1,631 @@
+            assert isinstance(output_el, ts.FieldType)
+
+        stencil_type_synthesizer = self.visit(node.stencil, ctx=ctx)
+        stencil_args = [
+            type_synthesizer._convert_as_fieldop_input_to_iterator(domain, input_)
+            for input_ in inputs
+        ]
+        stencil_returns = stencil_type_synthesizer(
+            *stencil_args, offset_provider=self.offset_provider
+        )
+
+        return it_ts.StencilClosureType(
+            domain=domain,
+            stencil=ts.FunctionType(
+                pos_only_args=stencil_args,
+                pos_or_kw_args={},
+                kw_only_args={},
+                returns=stencil_returns,
+            ),
+            output=output,
+            inputs=inputs,
+        )
+
+    def visit_AxisLiteral(self, node: itir.AxisLiteral, **kwargs) -> ts.DimensionType:
+        assert (
+            node.value in self.dimensions
+        ), f"Dimension {node.value} not present in offset provider."
+        return ts.DimensionType(dim=self.dimensions[node.value])
+
+    # TODO: revisit what we want to do with OffsetLiterals as we already have an Offset type in
+    #  the frontend.
+    def visit_OffsetLiteral(self, node: itir.OffsetLiteral, **kwargs) -> it_ts.OffsetLiteralType:
+        if _is_representable_as_int(node.value):
+            return it_ts.OffsetLiteralType(
+                value=ts.ScalarType(kind=getattr(ts.ScalarKind, itir.INTEGER_INDEX_BUILTIN.upper()))
+            )
+        else:
+            assert isinstance(node.value, str) and node.value in self.dimensions
+            return it_ts.OffsetLiteralType(value=self.dimensions[node.value])
+
+    def visit_Literal(self, node: itir.Literal, **kwargs) -> ts.ScalarType:
+        assert isinstance(node.type, ts.ScalarType)
+        return node.type
+
+    def visit_SymRef(
+        self, node: itir.SymRef, *, ctx: dict[str, ts.TypeSpec]
+    ) -> ts.TypeSpec | type_synthesizer.TypeSynthesizer:
+        # for testing, it is useful to be able to use types without a declaration
+        if self.allow_undeclared_symbols and node.id not in ctx:
+            # type has been stored in the node itself
+            if node.type:
+                if isinstance(node.type, ts.FunctionType):
+                    return _type_synthesizer_from_function_type(node.type)
+                return node.type
+            return ts.DeferredType(constraint=None)
+        assert node.id in ctx
+        result = ctx[node.id]
+        if isinstance(result, ObservableTypeSynthesizer):
+            result.aliases.append(node)
+        return result
+
+    def visit_Lambda(
+        self, node: itir.Lambda | itir.FunctionDefinition, *, ctx: dict[str, ts.TypeSpec]
+    ) -> ObservableTypeSynthesizer:
+        def fun(*args):
+            return self.visit(
+                node.expr, ctx=ctx | {p.id: a for p, a in zip(node.params, args, strict=True)}
+            )
+
+        return ObservableTypeSynthesizer(
```

Comment:
Shouldn't the `visit_Node()` postprocessing take care of converting the `TypeSynthesizer` into an `ObservableTypeSynthesizer` or am I missing somtehing?